django/db/models/sql/compiler.py

```diff
@@ -974,28 +974,39 @@ def _get_field_choices():
         result = []
         invalid_names = []
         for name in self.query.select_for_update_of:
-            parts = [] if name == 'self' else name.split(LOOKUP_SEP)
             klass_info = self.klass_info
-            for part in parts:
-                for related_klass_info in klass_info.get('related_klass_infos', []):
-                    field = related_klass_info['field']
-                    if related_klass_info['reverse']:
-                        field = field.remote_field
-                    if field.name == part:
-                        klass_info = related_klass_info
+            if name == 'self':
```

Comment:
We treat `self` as a separate case trying to find the first selected column from a queryset's model. We don't need to lock anything if such column doesn't exist. Previously we always took the first selected column in such cases which may not be related with the base model.